### PR TITLE
Design Sign-up

### PR DESCRIPTION
**Pull request by vuducdat**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -4,6 +4,8 @@ source "https://rubygems.org"
 # Bundle edge Rails instead: gem "rails", github: "rails/rails"
 gem "rails", "4.2.1"
 # Use sqlite3 as the database for Active Record
+gem 'bootstrap-sass', "3.2.0.0"
```
Comment (phonghenry14) [inline Gemfile]: "  "

File: app/assets/javascripts/users.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (phonghenry14) [inline app/assets/javascripts/users.coffee]: xóa file ko dùng

File: app/assets/stylesheets/users.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
```
Comment (phonghenry14) [inline app/assets/stylesheets/users.scss]: xóa

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+  def show
+    @user = User.find(params[:id])
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def new
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new(user_params) 
+    if @user.save
+      flash[:success] = "Welcome to the E-Learning!"
+      redirect_to @user
+    else
+      render 'new'
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: " "

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    end
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password,
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: bỏ () phía sau

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+      flash[:success] = "Welcome to the E-Learning!"
+      redirect_to @user
+    else
+      render 'new'
+    end
+  end
+
+  private
+
+    def user_params
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: viết liền vào private ko thụt đầu dòng

File: app/helpers/users_helper.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+module UsersHelper
+  def gravatar_for(user)
```
Comment (phonghenry14) [inline app/helpers/users_helper.rb]: bỏ ()

File: app/models/user.rb [hunk 10]
```diff
@@ -10,5 +10,12 @@ class User < ActiveRecord::Base
                                    dependent:   :destroy                                
   has_many :following, through: :active_relationships,  source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
-
+  before_save { self.email = email.downcase }
```
Comment (phonghenry14) [inline app/models/user.rb]: viết gọn {}

File: app/views/users/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<% provide(:title, 'Sign up') %>
```
Comment (phonghenry14) [inline app/views/users/new.html.erb]: bỏ () và " "

File: app/views/users/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+<h1>Sign up</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
+      <%= render 'shared/error_messages' %>
```
Comment (phonghenry14) [inline app/views/users/new.html.erb]: tất cả " "

File: app/views/users/show.html.erb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (phonghenry14) [inline app/views/users/show.html.erb]: bỏ ()

File: config/routes.rb [hunk 14]
```diff
@@ -1,9 +1,15 @@
 Rails.application.routes.draw do
+  get 'users/new'
```
Comment (phonghenry14) [inline config/routes.rb]: " "

File: db/migrate/20150616015348_add_index_to_users_email.rb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToUsersEmail < ActiveRecord::Migration
+  def change
+  	add_index :users, :email, unique: true
```
Comment (phonghenry14) [inline db/migrate/20150616015348_add_index_to_users_email.rb]: indent

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def new
+  	@user = User.new
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    @user = User.new(user_params) 
+    if @user.save
+      flash[:success] = "Welcome to the E-Learning!"
+      redirect_to @user
+    else
+      render 'new'
+    end
+  end
+
+  private
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]:
```ruby
private
def user_params
...
```

File: config/routes.rb [hunk 18]
```diff
@@ -1,9 +1,15 @@
...
+
+  get 'users/new'
```
Comment (VoTaiTri) [inline config/routes.rb]: remove this line

File: config/routes.rb [hunk 19]
```diff
@@ -1,9 +1,15 @@
...
+
   root "static_pages#home"
 
   get "help"      =>  "static_pages#help"
   get "about"     =>  "static_pages#about"
   get "contact"   =>  "static_pages#contact"
+  get 'signup'  => 'users#new'
+  resources :users
```
Comment (VoTaiTri) [inline config/routes.rb]:
```ruby
resources :users, only: [:show, :new, :create]
```

File: Gemfile [hunk 20]
```diff
@@ -4,6 +4,8 @@ source "https://rubygems.org"
 # Bundle edge Rails instead: gem "rails", github: "rails/rails"
 gem "rails", "4.2.1"
 # Use sqlite3 as the database for Active Record
+gem "bootstrap-sass", "3.2.0.0"
+gem 'bcrypt', "3.1.7"
```
Comment (leminhtuan2015) [inline Gemfile]: From now on please use ", not use '

Comment (phonghenry14) [inline Gemfile]: " " đây nữa e ơi

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,25 @@
+class UsersController < ApplicationController
+  def show
+    @user = User.find params[:id]
+  end
+
+  def new
+  	@user = User.new
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+      flash[:success] = "Welcome to the E-Learning!"
+      redirect_to @user
+    else
+      render "new"
+    end
+  end
+
+  private
+  def user_params
+    params.require(:user).permit :name,:email,:password,
```
Comment (phonghenry14) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb (hunk 21 above)
Comment (leminhtuan2015) [inline app/controllers/users_controller.rb]: Convention, please

File: config/database.yml.example [hunk 23]
```diff
@@ -1,25 +0,0 @@
-# SQLite version 3.x
```
Comment (phonghenry14) [inline config/database.yml.example]: ko xóa file này e ơi

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def new
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new user_params  
+    if @user.save
+      flash[:success] = "Welcome to the E-Learning!"
```
Comment (leminhtuan2015) [inline app/controllers/users_controller.rb]: It is better to use I18n for String, you should place "Welcome to the E-Learning!" in locale I18n
http://guides.rubyonrails.org/i18n.html

File: app/controllers/users_controller.rb [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+      redirect_to @user
+    else
+      render "new"
+    end
+  end
+
+  private
+  def user_params
+    params.require(:user).permit :name,:email,:password,
+                                   :password_confirmation
```
Comment (leminhtuan2015) [inline app/controllers/users_controller.rb]: Convention, please

File: config/routes.rb [hunk 26]
```diff
@@ -1,9 +1,12 @@
 Rails.application.routes.draw do
+  get "users/new"
+  get "users/new"
```
Comment (VoTaiTri) [inline config/routes.rb]: remove this line

File: app/helpers/users_helper.rb [hunk 27]
```diff
@@ -0,0 +1,7 @@
+module UsersHelper
+  def gravatar_for user
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
```
Comment (leminhtuan2015) [inline app/helpers/users_helper.rb]: Turn "https://secure.gravatar.com/avatar/" to Constant

File: app/models/user.rb [hunk 28]
```diff
@@ -1,7 +1,6 @@
 class User < ActiveRecord::Base
   has_many :lessons, dependent: :destroy
   has_many :activities, dependent: :destroy
-
   has_many :active_relationships,  class_name:  "Relationship",
                                    foreign_key: "follower_id",
                                    dependent:   :destroy
```
Comment (leminhtuan2015) [inline app/models/user.rb]: Convention, please

File: app/models/user.rb [hunk 29]
```diff
@@ -11,4 +10,13 @@ class User < ActiveRecord::Base
   has_many :following, through: :active_relationships,  source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
 
+  before_save {self.email = email.downcase}
+  
+  validates :name,  presence: true, length: {maximum: 50}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: 255},
+                    format: {with: VALID_EMAIL_REGEX},
+                    uniqueness: {case_sensitive: false}
```
Comment (leminhtuan2015) [inline app/models/user.rb]: Convention, please.

File: app/views/users/new.html.erb [hunk 30]
```diff
@@ -0,0 +1,24 @@
+<% provide :title, "Sign up" %>
+<h1>Sign up</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (leminhtuan2015) [inline app/views/users/new.html.erb]: form_for @user

File: config/routes.rb (hunk 26 above)
Comment (leminhtuan2015) [inline config/routes.rb]: Duplication

File: config/routes.rb [hunk 31]
```diff
@@ -1,9 +1,12 @@
...
   root "static_pages#home"
-
   get "help"      =>  "static_pages#help"
   get "about"     =>  "static_pages#about"
   get "contact"   =>  "static_pages#contact"
+  get "signup"  => "users#new"
+  resources :users, only:[:show,:new,:create]
```
Comment (leminhtuan2015) [inline config/routes.rb]: Space, please
only: [:show,:new,:create]

File: db/migrate/20150616015348_add_index_to_users_email.rb (hunk 15 above)
Comment (leminhtuan2015) [inline db/migrate/20150616015348_add_index_to_users_email.rb]: Convention, please

File: app/controllers/users_controller.rb [hunk 32]
```diff
@@ -0,0 +1,25 @@
...
+    @user = User.new
+  end
+
+  def create
+    @user = User.new user_params  
+    if @user.save
+      flash[:success] = t("title.welcome")
+      redirect_to @user
+    else
+      render "new"
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: should 
```ruby
render :new
``

File: app/models/user.rb [hunk 33]
```diff
@@ -1,14 +1,21 @@
...
   has_many :following, through: :active_relationships,  source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
 
+  before_save {self.email = email.downcase}
+  
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :name,  presence: true, length: {maximum: 50}
+  validates :email, presence: true, length: {maximum: 255},
+    format: {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false} 
+    has_secure_password
```
Comment (VoTaiTri) [inline app/models/user.rb]: indent

File: app/views/users/new.html.erb [hunk 34]
```diff
@@ -0,0 +1,24 @@
+<% provide :title, t("models.user") %>
+<h1>Sign up</h1>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/new.html.erb [hunk 35]
```diff
@@ -0,0 +1,24 @@
+<% provide :title, t("models.user") %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: rename I18n

File: app/controllers/users_controller.rb [hunk 36]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new user_params  
+    if @user.save
+      flash[:success] = t("title.welcome")
```
Comment (leminhtuan2015) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/models/user.rb [hunk 37]
```diff
@@ -1,14 +1,21 @@
 class User < ActiveRecord::Base
   has_many :lessons, dependent: :destroy
   has_many :activities, dependent: :destroy
-
   has_many :active_relationships,  class_name:  "Relationship",
-                                   foreign_key: "follower_id",
-                                   dependent:   :destroy
+    foreign_key: "follower_id",
+    dependent:   :destroy
```
Comment (leminhtuan2015) [inline app/models/user.rb]: Convention please

File: app/models/user.rb [hunk 38]
```diff
@@ -1,14 +1,21 @@
...
   has_many :passive_relationships, class_name:  "Relationship",
-                                   foreign_key: "followed_id",
-                                   dependent:   :destroy                                
+    foreign_key: "followed_id",
+    dependent:   :destroy
```
Comment (leminhtuan2015) [inline app/models/user.rb]: Convention please

File: app/models/user.rb [hunk 39]
```diff
@@ -1,14 +1,21 @@
...
   has_many :followers, through: :passive_relationships, source: :follower
 
+  before_save {self.email = email.downcase}
+  
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :name,  presence: true, length: {maximum: 50}
+  validates :email, presence: true, length: {maximum: 255},
+    format: {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false} 
+  has_secure_password
+  validates :password, presence: true, length: {minimum: 6}
```
Comment (leminhtuan2015) [inline app/models/user.rb]: You should group some code have the same role.
```ruby
VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
 
validates :name,  presence: true, length: {maximum: 50}
validates :email, presence: true, length: {maximum: 255},
validates :password, presence: true, length: {minimum: 6}

format: {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false} 

has_secure_password
```

File: app/views/layouts/_footer.html.erb [hunk 40]
```diff
@@ -1,6 +1,6 @@
 <footer class="footer">
   <small>
-    E-Learning Page by Tokyo
+    <%= t("title.footer") %>
```
Comment (leminhtuan2015) [inline app/views/layouts/_footer.html.erb]: Remove ( )

File: app/views/layouts/_header.html.erb [hunk 41]
```diff
@@ -1,6 +1,6 @@
 <header class="navbar navbar-fixed-top navbar-inverse">
   <div class="container">
-    <%= link_to "E-Learning Page", root_path, id: "logo" %>
+    <%= link_to t("title.header"), root_path, id: "logo" %>
     <nav>
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to "Home", root_path %></li>
```
Comment (leminhtuan2015) [inline app/views/layouts/_header.html.erb]: Should you place "Home" into I18n ???

File: app/views/shared/_error_messages.html.erb [hunk 42]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, "error") %>.
```
Comment (leminhtuan2015) [inline app/views/shared/_error_messages.html.erb]: I18n please

File: app/views/static_pages/help.html.erb [hunk 43]
```diff
@@ -1 +1 @@
-  <h1>Help</h1>
\ No newline at end of file
+  <h1><%= t("title.help") %></h1>
```
Comment (leminhtuan2015) [inline app/views/static_pages/help.html.erb]: Remove ( )

File: app/views/static_pages/home.html.erb [hunk 44]
```diff
@@ -1,4 +1,4 @@
 <div class="center jumbotron">
-  <h1>Welcome to the E-Learning Page</h1>
-  <%= link_to "Sign up now!", "#", class: "btn btn-lg btn-primary" %>
+  <h1><%= t("title.home") %></h1>
+  <%= link_to t("title.home1"), "#", class: "btn btn-lg btn-primary" %>
```
Comment (leminhtuan2015) [inline app/views/static_pages/home.html.erb]: Remove ( )

File: app/views/users/new.html.erb [hunk 45]
```diff
@@ -0,0 +1,24 @@
+<% provide :title, t("views.user") %>
+<h1><%= t("views.user") %></h1>
```
Comment (leminhtuan2015) [inline app/views/users/new.html.erb]: Remove ( )

File: app/views/users/new.html.erb [hunk 46]
```diff
@@ -0,0 +1,24 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+
+      <%= f.submit t("button.create"), class: "btn btn-primary" %>
```
Comment (leminhtuan2015) [inline app/views/users/new.html.erb]: Remove ( )

File: app/controllers/users_controller.rb [hunk 47]
```diff
@@ -0,0 +1,25 @@
...
+      redirect_to @user
+    else
+      render :new
+    end
+  end
+
+  private
+  def user_params
+    params.require(:user).permit :name, :email, :password, 
+    :password_confirmation
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent

File: app/helpers/users_helper.rb [hunk 48]
```diff
@@ -0,0 +1,8 @@
+module UsersHelper
+  GRAVATAR_URL = "https://secure.gravatar.com/avatar/#"
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: Store in Setting file

File: app/models/user.rb [hunk 49]
```diff
@@ -1,14 +1,21 @@
...
+    dependent: :destroy                                
   has_many :following, through: :active_relationships,  source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
 
+  before_save {self.email = email.downcase}
+  
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :name,  presence: true, length: {maximum: 50}
+  validates :email, presence: true, length: {maximum: 255},
+  format: {with: VALID_EMAIL_REGEX},uniqueness: {case_sensitive: false}
```
Comment (dieunb) [inline app/models/user.rb]: indent

File: app/views/users/new.html.erb [hunk 50]
```diff
@@ -0,0 +1,24 @@
+<% provide :title, t("views.user") %>
+<h1><%= t "views.user" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
```
Comment (dieunb) [inline app/views/users/new.html.erb]: Slice to new partial file with form

File: app/controllers/users_controller.rb (hunk 47 above)
Comment (leminhtuan2015) [inline app/controllers/users_controller.rb]:
```ruby
params.require(:user).permit :name, :email, :password,
  :password_confirmation
```

File: app/models/user.rb [hunk 51]
```diff
@@ -1,14 +1,21 @@
...
-                                   dependent:   :destroy                                
+    foreign_key: "followed_id",
+    dependent: :destroy                                
   has_many :following, through: :active_relationships,  source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
 
+  before_save {self.email = email.downcase}
+  
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :name,  presence: true, length: {maximum: 50}
```
Comment (leminhtuan2015) [inline app/models/user.rb]:
```ruby
validates :name, presence: true, length: {maximum: 50}
```

File: app/views/layouts/_header.html.erb [hunk 52]
```diff
@@ -1,11 +1,11 @@
...
-    <%= link_to "E-Learning Page", root_path, id: "logo" %>
+    <%= link_to t("title.header"), root_path, id: "logo" %>
     <nav>
       <ul class="nav navbar-nav navbar-right">
-        <li><%= link_to "Home", root_path %></li>
-        <li><%= link_to "Help", help_path %></li>
-        <li><%= link_to "Log in", '#' %></li>
+        <li><%= link_to t("title.home2"), root_path %></li>
+        <li><%= link_to t("title.help"), help_path %></li>
+        <li><%= link_to t("title.login"), '#' %></li>
```
Comment (leminhtuan2015) [inline app/views/layouts/_header.html.erb]: ' -> "

File: app/views/shared/_error_messages.html.erb [hunk 53]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t "title.error" %> <%= pluralize(@user.errors.count, "error") %>
```
Comment (leminhtuan2015) [inline app/views/shared/_error_messages.html.erb]: Space

File: config/routes.rb [hunk 54]
```diff
@@ -1,62 +1,9 @@
...
-  #   resources :posts, concerns: :toggleable
-  #   resources :photos, concerns: :toggleable
-
-  # Example resource route within a namespace:
-  #   namespace :admin do
-  #     # Directs /admin/products/* to Admin::ProductsController
-  #     # (app/controllers/admin/products_controller.rb)
-  #     resources :products
-  #   end
+  get "signup"  => "users#new"
```
Comment (leminhtuan2015) [inline config/routes.rb]: Space

**Pull request merged**